cpp/src/arrow/util/future.h

```diff
@@ -840,6 +840,18 @@ inline Future<>::Future(Status s) : Future(internal::Empty::ToResult(std::move(s
 ARROW_EXPORT
 Future<> AllComplete(const std::vector<Future<>>& futures);
 
+
+/// \brief Create a Future which completes when all of `futures` complete.
+///
+/// The future will be marked complete if all `futures` complete
+/// successfully. Otherwise, it will be marked failed with the status of
+/// the first failing future.
+///
+/// Unlike AllComplete this Future will not complete immediately when a
+/// failure occurs.  It will wait until all futures have finished.
+ARROW_EXPORT
+Future<> AllCompleteOrFailed(const std::vector<Future<>>& futures);
```

Comment:
nit: maybe `AllFinished`?

Reply:
this part of the code is more PR https://github.com/apache/arrow/pull/11017 which I needed to duplicate here in order to solve some race conditions here.